packages/core-transaction-pool/src/processor.ts

```diff
@@ -176,11 +172,28 @@ export class Processor implements TransactionPool.IProcessor {
             );
 
             return false;
-        } else if (transaction.expiration > 0 && transaction.expiration <= lastHeight + 1) {
+        }
+
+        const lastHeight: number = app
+            .resolvePlugin<State.IStateService>("state")
+            .getStore()
+            .getLastHeight();
+
+        const expirationContext = {
+            blockTime: Managers.configManager.getMilestone(lastHeight).blocktime,
+            currentHeight: lastHeight,
+            now: Crypto.Slots.getTime(),
+            maxTransactionAge: app.resolveOptions("transaction-pool").maxTransactionAge,
+        };
+
+        const transactionInstance: Interfaces.ITransaction = Transactions.TransactionFactory.fromData(transaction);
```

Comment:
We cannot do this. This call is not just expensive, but also the reason why the unit tests are blowing up now. I'd just move the expiration calculation into a utility function inside `core-transaction-pool` or `core-utils` since that doesn't really fit semantically into the transaction class in the first place.

Reply:
:ok_hand:

Reply:
> We cannot do this. This call is not just expensive, but also the reason why the unit tests are blowing up now.

The code before this patch already calls `fromData()`, next to the only invocation of `validateTransaction()`:

```
111 } else if (await this.validateTransaction(transaction)) {
112     try {
113         const receivedId: string = transaction.id;
114         const transactionInstance: Interfaces.ITransaction = Transactions.TransactionFactory.fromData(
115             transaction,
116         );
```

How could calling `fromData()` one more time break any test or be problematic in any way, other than performance wise? If that call is really expensive, then we can unite the 2 calls into a single one like this:

```
111 } else {    
112     const transactionInstance: Interfaces.ITransaction = Transactions.TransactionFactory.fromData(
113         transaction,
114     );  
115     if (await this.validateTransaction(transactionInstance)) {
```
and adjust `validateTransaction()` to take argument of type `ITransaction` instead of `ITransactionData`.

> the expiration calculation ... doesn't really fit semantically into the transaction class in the first place.

Why? Can you justify your reasoning?

I asked to get the expiration as a method of `ITransaction`. Here is my reasoning: the expiration is a property of a transaction, just like its id, amount, sender, or any of the other properties. Thus the expiration naturally belongs to the transaction class. But it cannot be a member variable because it is calculated differently for v1 and v2 transactions, so that naturally leads to having a method instead of a member variable, a method that does the calculation behind the scenes. It could even be a getter method (`get expiration()...`).

Reply:
> How could calling fromData() one more time break any test or be problematic in any way, other than performance wise? 

Because of a missing try catch.

> Why? Can you justify your reasoning?
Following that reasoning, we should also have signature calculation, id calculation and everything else part of the transaction class. While the calculated expiration is only needed by the transaction pool. Plus, what if the context is extended and at one point we need access to the container for it. Well, then it's out of scope for the transaction class for real.

Reply:
> > How could calling fromData() one more time break any test or be problematic in any way, other than performance wise?
> 
> Because of a missing try catch.

Well, then it can be added:
```
111 } else {
      try {    
112     const transactionInstance: Interfaces.ITransaction = Transactions.TransactionFactory.fromData(
113         transaction,
114     );
      } catch (...) {...}
115   if (await this.validateTransaction(transactionInstance)) {
```

> > Why? Can you justify your reasoning?
> 
> Following that reasoning, we should also have signature calculation, id calculation and everything else part of the transaction class.

Yes. You do `person.name()` not `calculatePersonName(person)`. I am not sure about "everything else". If some context is needed to get the name it is provided by the caller, e.g. `person.name(english)`, not `calculatePersonName(person, english)`. Also, this way if the calculation is expensive the class may cache it transparently and on subsequent calls to `name()` return the pre-calculated value.

> While the calculated expiration is only needed by the transaction pool.

Where it is used is irrelevant - if a property belongs to a class (e.g. name to a person, expiration to a transaction) it should be in that class. Also, where it is used from could change in the future.

> Plus, what if the context is extended and at one point we need access to the container for it. Well, then it's out of scope for the transaction class for real.

Reply:
I don't agree with `person.name(english)` either, since the person shouldn't have to care about  the localization. But I get what you are saying and agree that there are cases where it makes sense. However, I don't think this is such a case. It pollutes the Transaction class just to handle legacy transactions and it forces you to create a transaction instance which could be as well avoided.

So to come to an end... the PR looks good to me now and I prefer it over the previous approach. :+1: